### ISSUE TITLE
Allow serialising format_args! directly

### ISSUE THREAD

**Issue by clarcharr** [serde-rs/serde#786]
I've seen a fair number of `Serialize` implementations that involve calling `to_string` and I wonder if this extra allocation could be removed somehow. Take [this example](https://github.com/chronotope/chrono/blob/d036c0737059fe002463c10756553393bcbcc440/src/datetime.rs#L524) from chrono:

```
serializer.serialize_str(&format!("{:?}", self))
```

I wonder if we could perhaps modify this to:

```
serializer.serialize_str_format(format_args!("{:?}", self))
```

So that the serializer could call `write` itself directly to a buffer. For formats that don't require escaping characters, this could involve writing the data directly to the output stream, and for formats that do, this might involve writing to a temporary buffer (only allocated once) and then copying the data out with the escaped characters in it. Perhaps it could also default to calling `format` and then `serialize_str` if implementors choose to not specialise it.

Not sure how much of a performance gain this would be, but it seems worth investigating imo.

**Comment by clarcharr** [comment]
Possible specialisations for ease of use could be:

* `serialize_str_debug(val) -> serialize_str_format(format_args!("{:?}", val))`
* `serialize_str_display(val) -> serialize_str_format(format_args!("{:?}", val))`

**Comment by dtolnay** [comment]
Neat idea. The chrono impls could be optimized [similar to this](https://github.com/serde-rs/serde/pull/778) and I doubt there would be much performance left to gain after that, but it may make a difference in other cases. This ties in to #746 as well.

**Comment by dtolnay** [comment]
@nox would something like `Serializer::collect_str<T: Display>(self, T)` with a default impl do the trick?

**Comment by clarcharr** [comment]
Personally, I'd rather not limit this to just types with a `Display` impl because that requires a lot more boilerplate.

**Comment by nox** [comment]
@clarcharr This is the only way I can think of of using the Display trait while not making a breaking change. What do you suggest?

**Comment by dtolnay** [comment]
Can you give an example of the boilerplate it would require? FormatArgs has a Display impl so `serializer.collect_str(format_args!("{:?}", val))` would use the Debug impl of `val`.

**Comment by nox** [comment]
Or do you just mean to support `Debug` too?

**Comment by nox** [comment]
@dtolnay Actually it should probably be `Serializer::collect_str<T: Display>(self, &T)`, no?

**Comment by clarcharr** [comment]
Oh wait, I wasn't aware that `FormatArgs` implemented `Display`. That works.

**Comment by clarcharr** [comment]
The type of boilerplate I was thinking of was that using wrapper types to manually implement `Display` could feel a lot like boilerplate, but if `FormatArgs` allows that then I think that it's fine to only require that for complex implementations.

**Comment by dtolnay** [comment]
Instead of basing this on Display, it may be a good opportunity to base it on a trait of our own that is able to provide both a human-readable string representation and a compact binary representation. This would make a lot of sense for chrono, IpAddr etc. The idea would be that JSON and YAML and TOML serialize as a string while Bincode serializes as bytes.

**Comment by clarcharr** [comment]
I mean, I think that there's value in having a minimal display trait for strings as well. Because the value sent to `Display` is usually intended to be human-readable, there are often more compact ways to display things.

One example off the top of my head is that an object containing dates might want to display them in a format other than ISO, whereas a serialised format would likely prefer ISO.

**Comment by nox** [comment]
That seems like a job for `#[serde(with)]`, orthogonal to a method for the `Display` trait.

**Comment by dtolnay** [comment]
I filed #790 to follow up on the readable vs compact aspect of this in a separate change.

**Comment by dtolnay** [comment]
Fixed in #789.
